src/world/World.php

```diff
@@ -588,6 +602,193 @@ private function initRandomTickBlocksFromConfig(ServerConfigGroup $cfg) : void{
+
+		if($this->weatherTick >= $this->weatherDuration && $this->weatherDuration > 0){
+			$this->autoChangeWeather();
+		}
+
+		if($this->thunderLevel > 0.8 && mt_rand(0, 200) === 0){
+			$players = $this->getPlayers();
+			if(count($players) > 0){
+				$target = $players[array_rand($players)];
+				$spawn = $target->getPosition();
+				$x = (int) $spawn->x + mt_rand(-30, 30);
+				$z = (int) $spawn->z + mt_rand(-30, 30);
+				$y = $this->getHighestBlockAt($x, $z) ?? max(64, (int) $spawn->y);
+
+				$loc = Location::fromObject(new Vector3($x, $y, $z), $this);
+				(new LightningBolt($loc))->spawnToAll();
+			}
+		}
+	}
+
+	private function autoChangeWeather() : void{
+		$roll = mt_rand(1, 100);
+
+		if($roll <= 70){
+			$this->setWeather("clear", mt_rand(6000, 12000));
+		}elseif($roll <= 90){
+			$this->setWeather("rain", mt_rand(4000, 9000));
+		}else{
+			$this->setWeather("thunder", mt_rand(2000, 6000));
+		}
+	}
+
+	private function broadcastWeatherPackets() : void{
+		$players = $this->getPlayers();
+		if(count($players) === 0){
+			return;
+		}
+
+		$rainInt = (int) ($this->rainLevel * 65535);
+		$thunderInt = (int) ($this->thunderLevel * 65535);
+
+		$status = $this->getWeatherStatus();
+
+		switch($status){
+			case "clear":
+				NetworkBroadcastUtils::broadcastPackets($players, [
+					LevelEventPacket::create(LevelEvent::STOP_RAIN, 0, null),
+					LevelEventPacket::create(LevelEvent::STOP_THUNDER, 0, null),
+				]);
+				break;
+
+			case "rain":
+				NetworkBroadcastUtils::broadcastPackets($players, [
+					LevelEventPacket::create(LevelEvent::START_RAIN, $rainInt, null),
+					LevelEventPacket::create(LevelEvent::STOP_THUNDER, 0, null),
+				]);
+				break;
+
+			case "thunder":
+				NetworkBroadcastUtils::broadcastPackets($players, [
+					LevelEventPacket::create(LevelEvent::START_RAIN, $rainInt, null),
+					LevelEventPacket::create(LevelEvent::START_THUNDER, $thunderInt, null),
+				]);
+				break;
+		}
+	}
+
+	public function getRainLevel() : float{
+		return $this->rainLevel;
+	}
+
+	public function setRainLevel(float $level) : void {
+		if($level < 0.0 || $level > 1.0){
+			throw new \InvalidArgumentException("Rain level must be between 0.0 and 1.0, $level given");
+		}
+		$this->rainLevel = $level;
+	}
+
+	public function getThunderLevel() : float{
+		return $this->thunderLevel;
+	}
+
+	public function setThunderLevel(float $level) : void {
+		if($level < 0.0 || $level > 1.0){
+			throw new \InvalidArgumentException("Thunder level must be between 0.0 and 1.0, $level given");
+		}
+		$this->thunderLevel = $level;
+	}
+
+	public function isWeatherEnabled() : bool{
+		return $this->weatherEnabled;
+	}
+
+	public function isRainingAt(Vector3 $pos) : bool{
+		if($this->getRainLevel() <= 0){
+			return false;
+		}
+
+		$x = (int) $pos->x;
+		$y = (int) $pos->y;
+		$z = (int) $pos->z;
+
+		if(!$this->isChunkLoaded($x >> 4, $z >> 4)){
+			return false;
+		}
+
+		for($i = $y + 1, $maxY = $this->getMaxY(); $i <= $maxY; $i++){
```

Comment:
This is going to be laggy AF.
Use `World->getHighestBlockAt()` and count down until you hit a block that stops rain or you reach the affected Y coordinate.